staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -293,12 +293,14 @@ func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 		case event, ok := <-ch:
 			if !ok {
 				// End of results.
+				s.Watching.Stop()
 				return
 			}
 			obj := s.Fixup(event.Object)
 			if err := s.EmbeddedEncoder.Encode(obj, buf); err != nil {
 				// unexpected error
 				utilruntime.HandleError(fmt.Errorf("unable to encode watch object %T: %v", obj, err))
+				s.Watching.Stop()
```

Comment:
This seems correct. (disastrous if it happens, but correct)

Correctness changes should come with tests which demonstrate the problem they are correcting.

Reply:
Checked kube-apiserver.log from e2e-gce run.
Didn't see the log on line 304

Reply:
I think you need to reproduce this in a unit test. I don't think it will ever happen in e2e tests.

Reply:
Yeah.
Going over the tests in staging/src/k8s.io/apiserver/pkg/endpoints/handlers/ to see which test I can extrapolate.

Reply:
Here is how to reproduce the problem: queue up a big object to be transmitted, and have the receiver close the network connection during the transmission.

This leaks a goroutine if it happens, so I'm going to merge this ASAP. Can you please send a test as a followup, ideally a test that exercises every return statement here and proves that a goroutine doesn't leak?

Reply:
Let me try your suggestion in a followup.